bevy_reflect: Reflect remote types

MrGVSV opened the pull request:
# Objective

The goal with this PR is to allow the use of types that don't implement `Reflect` within the reflection API.

Rust's [orphan rule](https://doc.rust-lang.org/book/ch10-02-traits.html#implementing-a-trait-on-a-type) prevents implementing a trait on an external type when neither type nor trait are owned by the implementor. This means that if a crate, `cool_rust_lib`, defines a type, `Foo`, then a user cannot use it with reflection. What this means is that we have to ignore it most of the time:

```rust
#[derive(Reflect)]
struct SomeStruct {
  #[reflect(ignore)]
  data: cool_rust_lib::Foo
}
```

Obviously, it's impossible to implement `Reflect` on `Foo`. But does it *have* to be? 

Most of reflection doesn't deal with concrete types— it's almost all using `dyn Reflect`. And being very metadata-driven, it should theoretically be possible. I mean, [`serde`](https://serde.rs/remote-derive.html) does it.

## Solution

> Special thanks to @danielhenrymantilla for their help reviewing this PR and offering wisdom wrt safety.

Taking a page out of `serde`'s book, this PR adds the ability to easily use "remote types" with reflection. In this context, a "remote type" is the external type for which we have no ability to implement `Reflect`.

This adds the `#[reflect_remote(...)]` attribute macro, which is used to generate "remote type wrappers". All you have to do is define the wrapper exactly the same as the remote type's definition:

```rust
// Pretend this is our external crate
mod cool_rust_lib {
  #[derive(Default)]
  struct Foo {
    pub value: String
  }
}

#[reflect_remote(cool_rust_lib::Foo)]
struct FooWrapper {
  pub value: String
}
```

> **Note:** All fields in the external type *must* be public. This could be addressed with a separate getter/setter attribute either in this PR or in another one.

The macro takes this user-defined item and transforms it into a newtype wrapper around the external type, marking it as `#[repr(transparent)]`. The fields/variants defined by the user are simply used to build out the reflection impls.

Additionally, it generates an implementation of the new trait, `ReflectRemote`, which helps prevent accidental misuses of this API.

Therefore, the output generated by the macro would look something like:

```rust
#[repr(transparent)]
struct FooWrapper(pub cool_rust_lib::Foo);

impl ReflectRemote for FooWrapper {
  type Remote = cool_rust_lib::Foo;

  // transmutation methods...
}

// reflection impls...
// these will acknowledge and make use of the `value` field
```

Internally, the reflection API will pass around the `FooWrapper` and [transmute](https://doc.rust-lang.org/std/mem/fn.transmute.html) it where necessary. All we have to do is then tell `Reflect` to do that. So rather than ignoring the field, we tell `Reflect` to use our wrapper using the `#[reflect(remote = ...)]` field attribute:

```rust
#[derive(Reflect)]
struct SomeStruct {
  #[reflect(remote = FooWrapper)]
  data: cool_rust_lib::Foo
}
```

#### Other Macros & Type Data

Because this macro consumes the defined item and generates a new one, we can't just put our macros anywhere. All macros that should be passed to the generated struct need to come *below* this macro. For example, to derive `Default` and register its associated type data:

```rust
// ✅ GOOD
#[reflect_remote(cool_rust_lib::Foo)]
#[derive(Default)]
#[reflect(Default)]
struct FooWrapper {
  pub value: String
}

// ❌ BAD
#[derive(Default)]
#[reflect_remote(cool_rust_lib::Foo)]
#[reflect(Default)]
struct FooWrapper {
  pub value: String
}
```

#### Generics

Generics are forwarded to the generated struct as well. They should also be defined in the same order:

```rust
#[reflect_remote(RemoteGeneric<'a, T1, T2>)]
struct GenericWrapper<'a, T1, T2> {
  pub foo: &'a T1,
  pub bar: &'a T2,
}
```

> Naming does *not* need to match the original definition's. Only order matters here.

> Also note that the code above is just a demonstration and doesn't actually compile since we'd need to enforce certain bounds (e.g. `T1: Reflect`, `'a: 'static`, etc.)

#### Nesting

And, yes, you can nest remote types:

```rust
#[reflect_remote(RemoteOuter)]
struct OuterWrapper {
  #[reflect(remote = InnerWrapper)]
  pub inner: RemoteInner
}

#[reflect_remote(RemoteInner)]
struct InnerWrapper(usize);
```

#### Assertions

This macro will also generate some compile-time assertions to ensure that the correct types are used. It's important we catch this early so users don't have to wait for something to panic. And it also helps keep our `unsafe` a little safer.

For example, a wrapper definition that does not match its corresponding remote type will result in an error: 

```rust
mod external_crate {
  pub struct TheirStruct(pub u32);
}

#[reflect_remote(external_crate::TheirStruct)]
struct MyStruct(pub String); // ERROR: expected type `u32` but found `String`
```

<details>
<summary>Generated Assertion</summary>

```rust
const _: () = {
  #[allow(non_snake_case)]
  #[allow(unused_variables)]
  #[allow(unused_assignments)]
  #[allow(unreachable_patterns)]
  #[allow(clippy::multiple_bound_locations)]
  fn assert_wrapper_definition_matches_remote_type(
    mut __remote__: external_crate::TheirStruct,
  ) {
    __remote__.0 = (|| -> ::core::option::Option<String> { None })().unwrap();
  }
};
```

</details>

Additionally, using the incorrect type in a `#[reflect(remote = ...)]` attribute should result in an error:

```rust
mod external_crate {
  pub struct TheirFoo(pub u32);
  pub struct TheirBar(pub i32);
}

#[reflect_remote(external_crate::TheirFoo)]
struct MyFoo(pub u32);

#[reflect_remote(external_crate::TheirBar)]
struct MyBar(pub i32);

#[derive(Reflect)]
struct MyStruct {
  #[reflect(remote = MyBar)] // ERROR: expected type `TheirFoo` but found struct `TheirBar`
  foo: external_crate::TheirFoo
}
```

<details>
<summary>Generated Assertion</summary>

```rust
const _: () = {
    struct RemoteFieldAssertions;
    impl RemoteFieldAssertions {
        #[allow(non_snake_case)]
        #[allow(clippy::multiple_bound_locations)]
        fn assert__foo__is_valid_remote() {
            let _: <MyBar as bevy_reflect::ReflectRemote>::Remote = (|| -> ::core::option::Option<external_crate::TheirFoo> {
              None
            })().unwrap();
        }
    }
};
```

</details>

### Discussion

There are a couple points that I think still need discussion or validation.

- [x] 1. `Any` shenanigans

     ~~If we wanted to downcast our remote type from a `dyn Reflect`, we'd have to first downcast to the wrapper then extract the inner type. This PR has a [commit](b840db9f74cb6d357f951cb11b150d46bac89ee2) that addresses this by making all the `Reflect::*any` methods return the inner type rather than the wrapper type. This allows us to downcast directly to our remote type.~~

     ~~However, I'm not sure if this is something we want to do. For unknowing users, it could be confusing and seemingly inconsistent. Is it worth keeping? Or should this behavior be removed?~~

     I think this should be fine. The remote wrapper is an implementation detail and users should not need to downcast to the wrapper type. Feel free to let me know if there are other opinions on this though!

- [x] 2. Implementing `Deref/DerefMut` and `From`

     ~~We don't currently do this, but should we implement other traits on the generated transparent struct? We could implement `Deref`/`DerefMut` to easily access the inner type. And we could implement `From` for easier conversion between the two types (e.g. `T: Into<Foo>`).~~ As mentioned in the comments, we probably don't need to do this. Again, the remote wrapper is an implementation detail, and should generally not be used directly.
     
- [x] 3. ~~Should we define a getter/setter field attribute in this PR as well or leave it for a future one?~~ I think this should be saved for a future PR

- [ ] 4. Any foreseeable issues with this implementation?

#### Alternatives

One alternative to defining our own `ReflectRemote` would be to use [bytemuck's `TransparentWrapper`](https://docs.rs/bytemuck/1.13.1/bytemuck/trait.TransparentWrapper.html) (as suggested by @danielhenrymantilla).

This is definitely a viable option, as `ReflectRemote` is pretty much the same thing as `TransparentWrapper`. However, the cost would be bringing in a new crate— though, it is already in use in a few other sub-crates like bevy_render.

I think we're okay just defining `ReflectRemote` ourselves, but we can go the bytemuck route if we'd prefer offloading that work to another crate.

---

## Changelog

* Added the `#[reflect_remote(...)]` attribute macro to allow `Reflect` to be used on remote types
* Added `ReflectRemote` trait for ensuring proper remote wrapper usage

StarArawn commented:
This sounds really useful for any sort of editor or inspector in the future. Especially if a particular bevy plugin or non-bevy crate does not use the reflect types. I don't feel qualified to fully review this, as I don't know much about bevy's reflection stuff, but this PR is something that is needed!

ndarilek commented:
I definitely need this for [bevy_synthizer](https://labs.lightsout.games/projects/bevy_synthizer). Some of my types wrap theirs so I can use them as components. Unfortunately, for the moment I can't use any of these types in scenes, which is important since some of these components are audio-related properties that require specific tweaking per entity.

devil-ira commented:
It appears that nesting remote reflect with `FromReflect` fails.
Example:
```rust
use bevy::reflect::reflect_remote;

mod remote {
    pub struct A;

    pub struct B {
        pub a: A,
    }
}

#[reflect_remote(remote::A, FromReflect)]
struct LocalA;

#[reflect_remote(remote::B, FromReflect)] // Error: the trait `FromReflect` is not implemented for `remote::A`
struct LocalB {
    #[reflect(remote = "LocalA")]
    pub a: remote::A,
}
```

devil-ira commented:
Attributes on fields are lost when `#[reflect_remote]` is added.

```rust
mod remote {
    pub struct A {
        #[cfg(feature = "something")]
        pub a: i32,
    }
}

#[reflect_remote(remote::A)]
pub struct LocalA {
    #[cfg(feature = "something")]
    pub a: i32, // Error: no field `a` on type `remote::A`
}
```

MrGVSV commented:
@devil-ira The nesting issue should be resolved (as well as an issue with generics that I discovered while trying to fix it).

Unfortunately, I'm not sure how to go about fixing the attribute issue. It seems the best way to handle this (without adding complexity/specialized logic) would be to just add the attribute to the entire item:

```rust
mod remote {
    pub struct A {
        #[cfg(oof)]
        pub a: i32,
    }
}

#[cfg(not(oof))]
#[reflect_remote(remote::A)]
pub struct LocalA {}

#[cfg(oof)]
#[reflect_remote(remote::A)]
pub struct LocalA {
    pub a: i32,
}
```

I think if we want to fix this, it should probably happen in a followup PR.

MForster commented:
This feature would be very useful for displaying more information in `bevy-inspector-egui`. Currently a lot of very simple structs can't be debugged, just because I don't own them...

MrGVSV commented:
Rebased onto main. Also improved the generated assertions to output for all `Reflect` derives rather than just for `#[reflect_remote(...)]` attributes. This ensures the correct type is used in any reflected struct/enum that uses a remote type.

MrGVSV commented:
Added another assertion that ensures wrapper definitions match their remote types (see PR description for details).

MrGVSV commented:
Rebased.

Also, added a [commit](efbdd4a61ca3cf0c68cc829de94a72ce7f023261) that updates the assertion for `reflect_remote` so that we no longer display the types in "reverse" order (i.e. no longer `expected IncorrectType, found CorrectType`). It also now correctly attributes the assertion to the macro.

Let me know if I should revert this change. While it does make a bit more sense imo, it also makes the error output slightly more verbose (with some mangled hints). 

<details>
<summary><h3>Examples</summary>

#### Example 1

```rust
mod external_crate {
    #[derive(Debug, Default)]
    pub struct TheirType {
        pub value: String,
    }
}

#[reflect_remote(external_crate::TheirType)]
#[derive(Debug, Default)]
#[reflect(Debug, Default)]
struct MyType {
    pub value: Foo,
}

#[derive(Debug, Reflect)]
struct Foo;
```

##### Before

```
error[E0308]: mismatched types
    --> crates/bevy_reflect/src/lib.rs:1667:9
     |
1667 |         #[reflect_remote(external_crate::TheirType)]
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Foo`, found `String`
...
1671 |             pub value: Foo,
     |                        --- expected due to this
     |
     = note: this error originates in the attribute macro `reflect_remote` (in Nightly builds, run with -Z macro-backtrace for more info)
```

##### After
```
error[E0308]: mismatched types
    --> crates/bevy_reflect/src/lib.rs:1671:24
     |
1667 |         #[reflect_remote(external_crate::TheirType)]
     |         --------------------------------------------
     |         |
     |         expected due to the type of this binding
     |         in this procedural macro expansion
...
1671 |             pub value: Foo,
     |                        ^^^ expected `String`, found `Foo`
     |
     = note: this error originates in the attribute macro `reflect_remote` (in Nightly builds, run with -Z macro-backtrace for more info)
```

#### Example 2

```rust
mod external_crate {
    #[derive(Debug, Default)]
    pub struct TheirType {
        pub value: String,
    }
}

#[reflect_remote(external_crate::TheirType)]
#[derive(Debug, Default)]
#[reflect(Debug, Default)]
struct MyType {
    pub value: Foo,
}

#[derive(Debug, Reflect)]
struct Foo;

impl Into<String> for Foo {
    fn into(self) -> String {
        todo!()
    }
}
```

##### Before

```
error[E0308]: mismatched types
    --> crates/bevy_reflect/src/lib.rs:1667:9
     |
1667 |         #[reflect_remote(external_crate::TheirType)]
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Foo`, found `String`
...
1671 |             pub value: Foo,
     |                        --- expected due to this
     |
     = note: this error originates in the attribute macro `reflect_remote` (in Nightly builds, run with -Z macro-backtrace for more info)
```

##### After
```
error[E0308]: mismatched types
    --> crates/bevy_reflect/src/lib.rs:1671:24
     |
1667 |         #[reflect_remote(external_crate::TheirType)]
     |         --------------------------------------------
     |         |
     |         expected due to the type of this binding
     |         in this procedural macro expansion
...
1671 |             pub value: Foo,
     |                        ^^^ expected `String`, found `Foo`
     |
     = note: this error originates in the attribute macro `reflect_remote` (in Nightly builds, run with -Z macro-backtrace for more info)
help: call `Into::into` on this expression to convert `tests::foo::Foo` into `std::string::String`
     |
1671 |             pub value: Foo.into(),
     |                           +++++++
```

#### Example 3

```rust
mod external_crate {
    #[derive(Debug, Default)]
    pub enum TheirType {
        #[default]
        Unit,
        Tuple(usize),
        Struct {
            value: String,
        },
    }
}

#[reflect_remote(external_crate::TheirType, FromReflect)]
#[derive(Debug)]
#[reflect(Debug)]
enum MyType {
    Unit,
    Tuple(i32),
    Struct { value: f32 },
}
```

##### Before

```
error[E0308]: mismatched types
    --> crates/bevy_reflect/src/lib.rs:1713:9
     |
1713 |         #[reflect_remote(external_crate::TheirType, FromReflect)]
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `i32`, found `usize`
...
1718 |             Tuple(i32),
     |                   --- expected due to this
     |
     = note: this error originates in the attribute macro `reflect_remote` (in Nightly builds, run with -Z macro-backtrace for more info)
help: you can convert a `usize` to an `i32` and panic if the converted value doesn't fit
     |
1713 |         #[reflect_remote(external_crate::TheirType, FromReflect)].try_into().unwrap()
     |                                                                  ++++++++++++++++++++
```

```
error[E0308]: mismatched types
    --> crates/bevy_reflect/src/lib.rs:1713:9
     |
1713 |         #[reflect_remote(external_crate::TheirType, FromReflect)]
     |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `f32`, found `String`
...
1719 |             Struct { value: f32 },
     |                             --- expected due to this
     |
     = note: this error originates in the attribute macro `reflect_remote` (in Nightly builds, run with -Z macro-backtrace for more info)
```

##### After

```
error[E0308]: mismatched types
    --> crates/bevy_reflect/src/lib.rs:1718:19
     |
1713 |           #[reflect_remote(external_crate::TheirType, FromReflect)]
     |           ---------------------------------------------------------
     |           |
     |  _________in this procedural macro expansion
     | |
1714 | |         #[derive(Debug)]
1715 | |         #[reflect(Debug)]
1716 | |         enum MyType {
1717 | |             Unit,
1718 | |             Tuple(i32),
     | |                   ^^-
     | |___________________|_|
     |                     | expected due to the type of this binding
     |                     expected `usize`, found `i32`
     |
     = note: this error originates in the attribute macro `reflect_remote` (in Nightly builds, run with -Z macro-backtrace for more info)
```

```
error[E0308]: mismatched types
    --> crates/bevy_reflect/src/lib.rs:1719:29
     |
1713 |         #[reflect_remote(external_crate::TheirType, FromReflect)]
     |         --------------------------------------------------------- in this procedural macro expansion
...
1719 |             Struct { value: f32 },
     |                             ^^^- help: try using a conversion method: `.to_string()`
     |                             |
     |                             expected `String`, found `f32`
     |                             expected due to the type of this binding
     |
     = note: this error originates in the attribute macro `reflect_remote` (in Nightly builds, run with -Z macro-backtrace for more info)
```

</details>

MrGVSV commented:
Update: Added the unsafe `ReflectRemote` trait.

This trait is implemented automatically when using `#[reflect_remote]`, and indicates that the implementor is safe to use as a remote wrapper. The trait is then used inside a compile-time assertion.

This should help prevent people from doing something like:

```rust
#[derive(Reflect, Debug)]
struct ContainerStruct {
  #[reflect(remote = "InvalidWrapper")]
  their_type: external_crate::TheirType,
}

#[derive(Reflect, Debug)]
struct InvalidWrapper(#[reflect(ignore)] external_crate::TheirType);
```

We could use the [autoref specialization trick](https://github.com/dtolnay/case-studies/blob/master/autoref-specialization/README.md) to print a custom error message indicating to the user to use `#[reflect_remote]`. However, since the error points back to `ReflectRemote` and the docs for it say to use `#[reflect_remote]`, we probably don't need to do this.

danielhenrymantilla commented on crates/bevy_reflect/bevy_reflect_derive/src/from_reflect.rs:
```diff
@@ -234,6 +264,27 @@ fn get_active_fields(
+                if field.attrs.remote.is_some() {
+                    value = if field.attrs.is_remote_generic().unwrap_or_default() {
+                        quote! {
+                            (|| Some(
```
```suggestion
                            (|| #FQOption::Some(
```

danielhenrymantilla commented on crates/bevy_reflect/bevy_reflect_derive/src/from_reflect.rs:
```diff
@@ -234,6 +264,27 @@ fn get_active_fields(
+                        }
+                    } else {
+                        quote! {
+                            (|| Some(
```
```suggestion
                            (|| #FQOption::Some(
```

danielhenrymantilla commented on crates/bevy_reflect/bevy_reflect_derive/src/from_reflect.rs:
```diff
@@ -234,6 +264,27 @@ fn get_active_fields(
+                        }
+                    };
+                    value = quote! {
+                        (|| Some(
```
```suggestion
                        (|| #FQOption::Some(
```

danielhenrymantilla commented on crates/bevy_reflect/src/remote.rs:
```diff
@@ -0,0 +1,50 @@
+///
+/// [reflectable]: Reflect
+/// [`transmute`]: core::mem::transmute
+/// [highly unsafe]: https://doc.rust-lang.org/nomicon/transmutes.html
```
```suggestion
/// [very unsafe]: https://doc.rust-lang.org/nomicon/transmutes.html
```

(or use _highly_ above)

danielhenrymantilla replied:
```suggestion
/// [highly unsafe]: https://doc.rust-lang.org/1.70.0/nomicon/transmutes.html
```

permalink

danielhenrymantilla commented on crates/bevy_reflect/src/remote.rs:
```diff
@@ -0,0 +1,50 @@
+/// [`transmute`]: core::mem::transmute
+/// [highly unsafe]: https://doc.rust-lang.org/nomicon/transmutes.html
+/// [`FromReflect`]: crate::FromReflect
+pub unsafe trait ReflectRemote: Reflect {
```
Here is where I'd recommend using [`::bytemuck`'s `TransparentWrapper`](https://docs.rs/bytemuck/1.13.1/bytemuck/trait.TransparentWrapper.html):

  - it would let your trait not be `unsafe`;
  - it would give you the _targeted_ transmuting functions for free / without `unsafe`:

```suggestion
pub trait ReflectRemote: Reflect + ::bytemuck::TransparentWrapper<Self::Remote> {
```

MrGVSV commented:
Updated the PR to support Bevy 0.11.

This removes the need for the special `FromReflect` opt-in since now that `FromReflect` is derived by default.

I also updated `ReflectRemote`:
- It no longer is `unsafe` to implement
- It now defines methods related to remote and wrapper types (e.g. `into_remote`, `into_wrapper`, etc.).

The methods on `ReflectRemote` could likely have default implementations. However, I chose not to do this so as to make it harder for users to accidentally shoot themselves in the foot by doing something like:

```rust
#[derive(Reflect)]
struct NotSafeToMakeRemote {
  foo: i32
}

impl ReflectRemote for NotSafeToMakeRemote {
  type Remote = SomeRemoteType;
}
```

Testare reviewed (approved):
Looks pretty good to me! Love the extensive testing

MrGVSV commented on crates/bevy_reflect/bevy_reflect_derive/src/impls/enums.rs:
```diff
@@ -389,9 +427,16 @@ fn generate_impls(reflect_enum: &ReflectEnum, ref_index: &Ident, ref_name: &Iden
             EnumVariantFields::Unnamed(fields) => {
                 let args = get_field_args(fields, |reflect_idx, declaration_index, field| {
-                    let declare_field = syn::Index::from(declaration_index);
+                    let declar_field = syn::Index::from(declaration_index);
```
My guess is that this was a rebase slip-up (unless I intentionally just hated that `e` lol). I'll put it back!

MrGVSV commented on crates/bevy_reflect/bevy_reflect_derive/src/enum_utility.rs:
```diff
@@ -42,7 +41,9 @@ pub(crate) fn get_variant_constructors(
-            let field_ty = &field.data.ty;
+
+            // Use the remote type if it exists, otherwise default to the field's actual type.
+            let field_ty = field.attrs.remote.as_ref().unwrap_or(&field.data.ty);
```
Good catch!

ariofrio commented:
> 2. Implementing Deref/DerefMut and From
> 
> We don't currently do this, but should we implement other traits on the generated transparent struct? We could implement Deref/DerefMut to easily access the inner type. And we could implement From for easier conversion between the two types (e.g. T: Into<Foo>).

I don't see why most of these would be useful, except maybe for `FooWrapper::from(foo)`. From what I understand, the generated transparent struct is not meant to be used directly by users. The only place where it should be used is as the argument to `#[reflect(remote =`, and in the case described by (1) below. Otherwise, the user just sees the original struct everywhere else. Is my assumption incorrect?

> 1. Any shenanigans
> 
> If we wanted to downcast our remote type from a dyn Reflect, we'd have to first downcast to the wrapper then extract the inner type. This PR has a [commit](https://github.com/bevyengine/bevy/pull/b840db9f74cb6d357f951cb11b150d46bac89ee2) that addresses this by making all the Reflect::*any methods return the inner type rather than the wrapper type. This allows us to downcast directly to our remote type.
> 
> However, I'm not sure if this is something we want to do. For unknowing users, it could be confusing and seemingly inconsistent. Is it worth keeping? Or should this behavior be removed?

I don't think this is too confusing as long as the generated struct type doesn't implement Reflect, only ReflectRemote, similar to how `serde` [differentiates them](https://serde.rs/remote-derive.html#invoking-the-remote-impl-directly):

> The code ultimately generated by these remote derives are not Serialize and Deserialize impls but associated functions with the same signature.

So the rule would be: Methods on a Reflect type return itself. Methods on ReflectRemote type return the inner type.

As far as converting between the two types, it seems to only go one way: you need to call `FooWrapper(foo).as_any()` (or `FooWrapper::from(foo).as_any()` or `FooWrapper::wrap(foo).as_any()`) but you never have a FooWrapper lying around that you need to get Foo out of (so you never need to call `Foo::from(foo_wrapper)`). Again, is my understanding of how this would be used correct? I quite possibly am forgetting about important use cases, not being deeply familiar with bevy_reflect.

Passing around FooWrapper rather than doing the conversion only when needed would probably be bad style, but there may be extreme cases where the user needs to do it anyway, though I can't think of any right now. And they would still be able to use `.0`, `peel`, or `peel_mut` (or whatever the equivalent would be for `ReflectRemote` if not using `TransparentWrapper`. This, btw, is a weak argument for using `TransparentWrapper`, not for its implementation, but for its well-understood and well-documented semantics.

soqb commented on crates/bevy_reflect/src/remote.rs:
```diff
@@ -0,0 +1,64 @@
+    fn into_remote(self) -> Self::Remote;
+
+    /// Converts a reference of the remote type to a reference of this wrapper.
+    fn as_wrapper(remote: &Self::Remote) -> &Self;
```
nit: i think i prefer `from_remote_ref` etc. for these.

MrGVSV replied:
Hm, I guess that does help keep things focused on the remote type. The only downside is the reduced parallelism with the other methods (i.e. `as_remote`/`as_wrapper` becomes `as_remote`/`from_remote_ref`).

Or did you mean to change them all to be `into_***`/`from_***`? That would be a bit nicer in terms of parallelism, but then it might seem counterintuitive that the `into_remote_ref` and `into_remote_mut` methods don't actually consume `self`.

So personally, I prefer the current method names the best, but I'm open to changing them if we feel that it would be better/clearer that way.

Thoughts?

soqb replied:
this is a really minor nit so i don't think my tiny request for consistency is really worth making any change here.

soqb reviewed (approved):
i'm so sorry, i thought i had reviewed this a long time ago!

generally excellent but i have a few nits/questions.

MrGVSV commented on crates/bevy_reflect/bevy_reflect_derive/src/derive_data.rs:
```diff
@@ -954,6 +1051,28 @@ impl<'a> ReflectTypePath<'a> {
+    /// To get the custom path if there is one, use [`Self::get_path`].
+    ///
+    /// For example, the type `Foo<T: Debug>` would return `Foo<T>`.
+    pub fn true_type(&self) -> Type {
```
I doubt we _need_ `Type`. I'll try to make the swap to `TokenStream`!

MrGVSV commented on crates/bevy_reflect/bevy_reflect_derive/src/field_attributes.rs:
```diff
@@ -176,4 +182,42 @@ impl FieldAttributes {
+        input.parse::<kw::remote>()?;
+        input.parse::<Token![=]>()?;
+
+        let lit = input.parse::<LitStr>()?;
+        self.remote = Some(lit.parse()?);
```
Me too. I'm guessing I chose this to be "consistent" with other reflection attributes (which is pretty much just `#[reflect(default)]` haha). I'll change it to take the non-stringified type path!

MrGVSV commented on crates/bevy_reflect/bevy_reflect_derive/src/from_reflect.rs:
```diff
@@ -191,34 +221,76 @@ fn get_active_fields(
+                            #FQOption::Some(
+                                // SAFETY: The remote type should always be a `#[repr(transparent)]` for the actual field type
+                                unsafe {
+                                    ::core::mem::transmute_copy::<#ty, #real_ty>(
+                                        &::core::mem::ManuallyDrop::new(#value?)
+                                    )
```
It's been a while so I don't quite remember 😅 

I'll test it out again and leave a comment above this line based on my findings. I think it had to do with lifetimes or something, but yeah I'll check again.

soqb replied:
yeah that's what i was worried about. i'm not 100% sure but i'm always wary of `transmute_copy` especially because it can unsoundly extend/shorten lifetimes!

MrGVSV replied:
Updated with a comment. Turns out it had to do with the compiler being unable to determine that generic types have the same size.

MrGVSV commented on crates/bevy_reflect/bevy_reflect_derive/src/impls/any.rs:
```diff
@@ -0,0 +1,43 @@
+use bevy_macro_utils::fq_std::{FQAny, FQBox};
+use quote::quote;
+
+/// A helper function for quickly implementing the `Reflect::*_any` methods while taking remote types into account.
```
Great minds think alike!

MrGVSV commented on crates/bevy_reflect/bevy_reflect_derive/src/remote.rs:
```diff
@@ -0,0 +1,474 @@
+    quote! {
+        #(#attrs)*
+        #[repr(transparent)]
+        #vis struct #ident #ty_generics (pub #remote_ty) #where_clause;
```
We don't have to, since it is pretty much an implementation detail. But I included it so that users could have the choice to pass around the wrapper type if they so choose.

However, I can remove it for now so users don't accidentally pollute their public API. And we can add it back in if it ever comes up as a need for someone.

soqb replied:
there's already `as_wrapper` etc. for this. i just don't think we need to commit to this implementation as part of the public api.

MrGVSV replied:
As it turns out, this is necessary. 😅

The reason is because we might define our remote type in one module, but wish to use it from another:

```rust
mod wrapper {
    #[reflect_remote(external_crate::TheirType)]
    pub struct MyType {
        pub value: String,
    }
}

mod foo {
    #[derive(Reflect)]
    struct ContainerStruct {
        #[reflect(remote = super::wrapper::MyType)]
        their_type: external_crate::TheirType,
    }
}
```

And we can't simply determine it from `TheirType` since `ReflectRemote` is only implemented for `MyType`.

MrGVSV commented:
> This, btw, is a weak argument for using `TransparentWrapper`, not for its implementation, but for its well-understood and well-documented semantics.

Yeah, I'm thinking it may be worth looking into using `TransparentWrapper` directly. It doesn't really do anything we don't already do here, but it has better safety assertions/documentation.

We could also just copy+paste the specific parts we need (since we really only need these six methods).

Whether we decide to use `bytemuck` or not, I think it's best to explore that in a future PR so that this one isn't further blocked/delayed more than it already has been.

MrGVSV commented:
I once again marked `ReflectRemote` as `unsafe` because I noticed that `bytemuck`'s `TransparentWrapper` is also marked as `unsafe`. Until we explore using `TransparentWrapper` directly in a future PR, it's probably best to at least align with it there.

soqb commented:
`TransparentWrapper` is unsafe because it explicitly guarantees that every type implementing it is marked with `#[repr(transparent)]`. there's no way to prove that in the system, so it's `unsafe` because implementors must manually assert that the implementee has a valid `#[repr(transparent)]` annotation.

that's not needed for `ReflectRemote`. it would be perfectly sound to implement `ReflectRemote` in any and every way, excluding of course where the method bodies are unsound in isolation which isn't grounds for making the trait `unsafe`. there's nothing we need to manually assert for `ReflectRemote` and relevant systems to be sound, just that weird implementations might produce logic errors.

`RemoteReflect` is more like [`ref_cast::RefCast`](https://docs.rs/ref-cast/latest/ref_cast/trait.RefCast.html) than `bytemuck::TransparentWrapper`

MrGVSV commented:
> `TransparentWrapper` is unsafe because it explicitly guarantees that every type implementing it is marked with `#[repr(transparent)]`. there's no way to prove that in the system, so it's `unsafe` because implementors must manually assert that the implementee has a valid `#[repr(transparent)]` annotation.

Oh that's true. It would only be "unsafe" if we provided the methods with the transmutations. But since those are only generated along with `repr(transparent)`, it should be perfectly fine.

Am I understanding that right?

If so, I can undo that last commit (again lol).

soqb commented:
yep that's all correct!

alice-i-cecile reviewed (approved):
As expected: thoroughly tested, well documented and solidly engineered.

The PR description was very helpful in understanding the how and why of this PR: while simply reading through the files it was easy to get bogged down in the details. I'm comfortable with the core mechanism here, the motivation is *very* strong, and the thorough reviews on safety were very useful.

alice-i-cecile commented:
I like the last minute reflect remote value addition! Let's get this merged :)

alice-i-cecile commented:
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1674 if you'd like to help out.
